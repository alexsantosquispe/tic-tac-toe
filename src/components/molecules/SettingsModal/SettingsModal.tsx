import { useTranslation } from 'react-i18next';
import useGame from '../../../hooks/useGame';
import { PLAYER_MODE, type OptionType } from '../../../models/types';
import Button from '../../atoms/Button/Button';
import { Modal } from '../../atoms/Modal/Modal';
import PlayersCard from '../../atoms/PlayersCard/PlayersCard';
import { PlayerModeOptions } from './components/PlayerModeOptions/PlayerModeOptions';

interface SettingsModalProps {
  isOpen: boolean;
  onClose: () => void;
}

const options: OptionType[] = [
  {
    value: PLAYER_MODE.SINGLE_PLAYER,
    component: <PlayersCard />
  },
  {
    value: PLAYER_MODE.TWO_PLAYERS,
    component: <PlayersCard isSinglePlayer={false} />
  }
];

const SettingsModal = ({ isOpen, onClose }: SettingsModalProps) => {
  const { t } = useTranslation();
  const { playerMode, setPlayerMode } = useGame();

  const onSelectOption = (value: OptionType['value']) => {
    setPlayerMode(value);
  };

  const saveSettings = () => {
    onClose();
  };

  return (
    <Modal title={t('settings.title')} isOpen={isOpen} onClose={onClose}>
      <div className="flex w-full flex-col gap-3 px-6 py-6">
        <span className="text-center">{t('settings.description')}</span>

        <PlayerModeOptions
          options={options}
          initialOptionValue={playerMode}
          onSelectOption={onSelectOption}
        />

        <Button
          title={t('settings.continueButton')}
          ariaLabel={t('settings.continueButtonAriaLabel')}
          onClick={saveSettings}
        />
      </div>
    </Modal>
  );
};

export default SettingsModal;
