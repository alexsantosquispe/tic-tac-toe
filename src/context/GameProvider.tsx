import { useState, type ReactNode } from 'react';
import { DEFAULT_DATA } from '../constants';
import type { SquareValueTypes } from '../types';
import { getWinnerResult } from '../utils';
import GameContext from './GameContext';

interface GameProviderProps {
  children: ReactNode;
}

const GameProvider = ({ children }: GameProviderProps) => {
  const [currentPlayer, setCurrentPlayer] = useState<'X' | 'O'>('X');
  const [data, setData] = useState<SquareValueTypes[]>(DEFAULT_DATA);
  const [winner, setWinner] = useState<SquareValueTypes | null>(null);
  const [combHighLight, setCombHighLight] = useState<number[]>([]);

  const toggleCurrentPlayer = () => {
    setCurrentPlayer((prevPlayer) => (prevPlayer === 'X' ? 'O' : 'X'));
  };

  const checkSquare = (index: number) => {
    if (data[index] !== '' || winner) return;
    const newData = [...data];
    newData[index] = currentPlayer;
    setData(newData);
    const result = getWinnerResult(newData);
    if (!result.winner) {
      toggleCurrentPlayer();
    }
    setWinner(result.winner);
    setCombHighLight(result.combHighLight);
  };

  const resetGame = () => {
    setData(DEFAULT_DATA);
    setWinner(null);
    setCombHighLight([]);
    setCurrentPlayer('X');
  };

  return (
    <GameContext.Provider
      value={{
        data,
        currentPlayer,
        winner,
        combHighLight,
        checkSquare,
        toggleCurrentPlayer,
        resetGame
      }}
    >
      {children}
    </GameContext.Provider>
  );
};

export default GameProvider;
