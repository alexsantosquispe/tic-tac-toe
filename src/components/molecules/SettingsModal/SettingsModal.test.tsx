import '@testing-library/jest-dom';

import { fireEvent, render, screen } from '@testing-library/react';

import GameProvider from '../../../context/GameProvider';
import { I18NWrapper } from '../../../tests/testsUtils';
import SettingsModal from './SettingsModal';

describe('SettingsModal', () => {
  const onCloseMock = jest.fn();
  const props = {
    isOpen: true,
    onClose: onCloseMock
  };

  it('should render the component correctly', () => {
    const component = render(
      <I18NWrapper>
        <GameProvider>
          <SettingsModal {...props} />
        </GameProvider>
      </I18NWrapper>
    );

    expect(component).toMatchSnapshot();
  });

  describe('behavior', () => {
    beforeEach(() => {
      render(
        <GameProvider>
          <SettingsModal {...props} />
        </GameProvider>
      );
    });
    it('should call onClose when the close button is clicked', () => {
      const closeButton = screen.getByRole('button', {
        name: 'Close modal button'
      });

      expect(closeButton).toBeInTheDocument();

      fireEvent.click(closeButton);

      expect(onCloseMock).toHaveBeenCalledTimes(1);
    });

    it('should check the player mode singlePlayer and select twoPlayers', () => {
      const singlePlayerButton = screen.getByRole('button', {
        name: 'Single player button'
      });

      const twoPlayersButton = screen.getByRole('button', {
        name: 'Two players button'
      });

      expect(singlePlayerButton).toBeInTheDocument();

      expect(twoPlayersButton).toBeInTheDocument();

      expect(singlePlayerButton).toHaveClass(
        'border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70'
      );

      fireEvent.click(twoPlayersButton);

      expect(twoPlayersButton).toHaveClass(
        'border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70'
      );
    });
  });
});
