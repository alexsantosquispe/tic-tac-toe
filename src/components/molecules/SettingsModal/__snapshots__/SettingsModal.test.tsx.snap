// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`SettingsModal should render the component correctly 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="fixed z-50 flex h-full w-full items-center justify-center bg-black/50 p-8 backdrop-blur-xs dark:bg-black/20"
        data-testid="modal"
        style="opacity: 0;"
      >
        <div
          class="flex w-full flex-col rounded-2xl bg-white md:w-80 dark:border dark:border-white/20 dark:bg-neutral-950"
          style="opacity: 0; transform: translateY(20px) scale(0.9);"
        >
          <button
            aria-label="Close modal button"
            class="mt-5 mr-5 self-end rounded-xl text-neutral-500 hover:cursor-pointer hover:text-black dark:text-neutral-600 dark:hover:text-white"
          >
            <svg
              class="size-6"
              data-testid="x-icon"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="3"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 6 6 18"
              />
              <path
                d="m6 6 12 12"
              />
            </svg>
          </button>
          <h2
            class="text-center text-xl font-medium md:text-2xl"
          >
            Settings
          </h2>
          <div
            class="flex w-full"
          >
            <div
              class="flex w-full flex-col gap-3 px-6 py-6"
            >
              <span
                class="text-center"
              >
                Choose your gaming mode
              </span>
              <div
                class="flex flex-col gap-3"
              >
                <button
                  aria-label="Single player button"
                  class="rounded-xl border-2 hover:cursor-pointer border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70"
                >
                  <div
                    class="flex w-full items-center justify-between gap-8 px-8 py-6"
                  >
                    <svg
                      class="size-12"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                      />
                      <circle
                        cx="12"
                        cy="7"
                        r="4"
                      />
                    </svg>
                    <span
                      class="text-xl font-medium"
                    >
                      VS
                    </span>
                    <svg
                      class="size-12"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M12 8V4H8"
                      />
                      <rect
                        height="12"
                        rx="2"
                        width="16"
                        x="4"
                        y="8"
                      />
                      <path
                        d="M2 14h2"
                      />
                      <path
                        d="M20 14h2"
                      />
                      <path
                        d="M15 13v2"
                      />
                      <path
                        d="M9 13v2"
                      />
                    </svg>
                  </div>
                </button>
                <button
                  aria-label="Two players button"
                  class="rounded-xl border-2 border-neutral-300 hover:cursor-pointer dark:border-white/20"
                >
                  <div
                    class="flex w-full items-center justify-between gap-8 px-8 py-6"
                  >
                    <svg
                      class="size-12"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                      />
                      <circle
                        cx="12"
                        cy="7"
                        r="4"
                      />
                    </svg>
                    <span
                      class="text-xl font-medium"
                    >
                      VS
                    </span>
                    <svg
                      class="size-12"
                      fill="none"
                      height="24"
                      stroke="currentColor"
                      stroke-linecap="round"
                      stroke-linejoin="round"
                      stroke-width="2"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                      />
                      <circle
                        cx="12"
                        cy="7"
                        r="4"
                      />
                    </svg>
                  </div>
                </button>
              </div>
            </div>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="fixed z-50 flex h-full w-full items-center justify-center bg-black/50 p-8 backdrop-blur-xs dark:bg-black/20"
      data-testid="modal"
      style="opacity: 0;"
    >
      <div
        class="flex w-full flex-col rounded-2xl bg-white md:w-80 dark:border dark:border-white/20 dark:bg-neutral-950"
        style="opacity: 0; transform: translateY(20px) scale(0.9);"
      >
        <button
          aria-label="Close modal button"
          class="mt-5 mr-5 self-end rounded-xl text-neutral-500 hover:cursor-pointer hover:text-black dark:text-neutral-600 dark:hover:text-white"
        >
          <svg
            class="size-6"
            data-testid="x-icon"
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="3"
            viewBox="0 0 24 24"
            width="24"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18 6 6 18"
            />
            <path
              d="m6 6 12 12"
            />
          </svg>
        </button>
        <h2
          class="text-center text-xl font-medium md:text-2xl"
        >
          Settings
        </h2>
        <div
          class="flex w-full"
        >
          <div
            class="flex w-full flex-col gap-3 px-6 py-6"
          >
            <span
              class="text-center"
            >
              Choose your gaming mode
            </span>
            <div
              class="flex flex-col gap-3"
            >
              <button
                aria-label="Single player button"
                class="rounded-xl border-2 hover:cursor-pointer border-rose-600 bg-rose-100 dark:border-rose-600 dark:bg-rose-950/70"
              >
                <div
                  class="flex w-full items-center justify-between gap-8 px-8 py-6"
                >
                  <svg
                    class="size-12"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                    />
                    <circle
                      cx="12"
                      cy="7"
                      r="4"
                    />
                  </svg>
                  <span
                    class="text-xl font-medium"
                  >
                    VS
                  </span>
                  <svg
                    class="size-12"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M12 8V4H8"
                    />
                    <rect
                      height="12"
                      rx="2"
                      width="16"
                      x="4"
                      y="8"
                    />
                    <path
                      d="M2 14h2"
                    />
                    <path
                      d="M20 14h2"
                    />
                    <path
                      d="M15 13v2"
                    />
                    <path
                      d="M9 13v2"
                    />
                  </svg>
                </div>
              </button>
              <button
                aria-label="Two players button"
                class="rounded-xl border-2 border-neutral-300 hover:cursor-pointer dark:border-white/20"
              >
                <div
                  class="flex w-full items-center justify-between gap-8 px-8 py-6"
                >
                  <svg
                    class="size-12"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                    />
                    <circle
                      cx="12"
                      cy="7"
                      r="4"
                    />
                  </svg>
                  <span
                    class="text-xl font-medium"
                  >
                    VS
                  </span>
                  <svg
                    class="size-12"
                    fill="none"
                    height="24"
                    stroke="currentColor"
                    stroke-linecap="round"
                    stroke-linejoin="round"
                    stroke-width="2"
                    viewBox="0 0 24 24"
                    width="24"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
                    />
                    <circle
                      cx="12"
                      cy="7"
                      r="4"
                    />
                  </svg>
                </div>
              </button>
            </div>
          </div>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
